docs: testing guide for CLI

wardbell opened the pull request:
Update testing guide to reflect CLI usage and current "best practices"
See #19510

updates tests
- heavy prose revisions
- uses `HttpClient` (with angular-in-memory-web-api)
- test `HeroService` using `HttpClientTestingModule`
- scrub away most `By.css`
- fake async observable with `asyncData()`
- extensive `TwainComponent` work
- different take on `retryWhen`
- remove app barrels (& `systemjs.extras`) which troubled plunker/systemjs
- add dummy `export const` to `hero.ts` (plunkr/systemjs fails w/o it)
- shrink and re-organize TOC
- add marble testing package and tests
- demonstrate the "no `beforeEach()`" test coding style
- add section on HTTP service testing
- prepare for stackblitz PR #20165
- make it work in plunker except excluded marble test

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes 
[ ] Other... Please describe:
```

wardbell committed:
docs: testing guide for CLI

- updates tests
- heavy prose revisions
- uses HttpClient (with angular-in-memory-web-api)
- test HeroService using `HttpClientTestingModule`
- scrub away most By.CSS
- fake async observable with `asyncData()`
- extensive Twain work
- different take on retryWhen
- remove app barrels (& systemjs.extras) which troubled plunker/systemjs
- add dummy export const to hero.ts (plunkr/systemjs fails w/o it)
- shrink and re-organize TOC
- add marble testing package and tests
- demonstrate the "no beforeEach()" test coding style
- add section on Http service testing
- prepare for stackblitz
- confirm works in plunker except excluded marble test
- add tests for avoidFile class feature of CodeExampleComponent

mary-poppins commented:
You can preview a14e2b6 at https://pr20697-a14e2b6.ngbuilds.io/.

mary-poppins commented:
You can preview 1cb3b2f at https://pr20697-1cb3b2f.ngbuilds.io/.

mary-poppins commented:
You can preview 7159fef at https://pr20697-7159fef.ngbuilds.io/.

mary-poppins commented:
You can preview 8f020a5 at https://pr20697-8f020a5.ngbuilds.io/.

mary-poppins commented:
You can preview b831b5a at https://pr20697-b831b5a.ngbuilds.io/.

mary-poppins commented:
You can preview 822245d at https://pr20697-822245d.ngbuilds.io/.

mary-poppins commented:
You can preview 2552d12 at https://pr20697-2552d12.ngbuilds.io/.

mary-poppins commented:
You can preview 28edf58 at https://pr20697-28edf58.ngbuilds.io/.

mary-poppins commented:
You can preview 7c3d378 at https://pr20697-7c3d378.ngbuilds.io/.

mary-poppins commented:
You can preview 500f414 at https://pr20697-500f414.ngbuilds.io/.

mary-poppins commented:
You can preview f3409ab at https://pr20697-f3409ab.ngbuilds.io/.

mary-poppins commented:
You can preview eb07c9d at https://pr20697-eb07c9d.ngbuilds.io/.

mary-poppins commented:
You can preview 7fa6a01 at https://pr20697-7fa6a01.ngbuilds.io/.

mary-poppins commented:
You can preview 30407c6 at https://pr20697-30407c6.ngbuilds.io/.

mary-poppins commented:
You can preview f4f0b6e at https://pr20697-f4f0b6e.ngbuilds.io/.

mary-poppins commented:
You can preview 9165af7 at https://pr20697-9165af7.ngbuilds.io/.

mary-poppins commented:
You can preview c4c97d3 at https://pr20697-c4c97d3.ngbuilds.io/.

mary-poppins commented:
You can preview 6b46e37 at https://pr20697-6b46e37.ngbuilds.io/.

mary-poppins commented:
You can preview a14e2b6 at https://pr20697-a14e2b6.ngbuilds.io/.
You can preview 1cb3b2f at https://pr20697-1cb3b2f.ngbuilds.io/.
You can preview 7159fef at https://pr20697-7159fef.ngbuilds.io/.
You can preview 8f020a5 at https://pr20697-8f020a5.ngbuilds.io/.
You can preview b831b5a at https://pr20697-b831b5a.ngbuilds.io/.
You can preview 822245d at https://pr20697-822245d.ngbuilds.io/.
You can preview 2552d12 at https://pr20697-2552d12.ngbuilds.io/.
You can preview 28edf58 at https://pr20697-28edf58.ngbuilds.io/.
You can preview 7c3d378 at https://pr20697-7c3d378.ngbuilds.io/.
You can preview 500f414 at https://pr20697-500f414.ngbuilds.io/.
You can preview f3409ab at https://pr20697-f3409ab.ngbuilds.io/.
You can preview eb07c9d at https://pr20697-eb07c9d.ngbuilds.io/.
You can preview 7fa6a01 at https://pr20697-7fa6a01.ngbuilds.io/.
You can preview 30407c6 at https://pr20697-30407c6.ngbuilds.io/.
You can preview f4f0b6e at https://pr20697-f4f0b6e.ngbuilds.io/.
You can preview 9165af7 at https://pr20697-9165af7.ngbuilds.io/.
You can preview c4c97d3 at https://pr20697-c4c97d3.ngbuilds.io/.
You can preview 6b46e37 at https://pr20697-6b46e37.ngbuilds.io/.
You can preview 5967dd1 at https://pr20697-5967dd1.ngbuilds.io/.

mary-poppins commented:
You can preview 8a546ca at https://pr20697-8a546ca.ngbuilds.io/.

mary-poppins commented:
You can preview cc09fdd at https://pr20697-cc09fdd.ngbuilds.io/.

mary-poppins commented:
You can preview 772f77b at https://pr20697-772f77b.ngbuilds.io/.

mary-poppins commented:
You can preview fcd5fb9 at https://pr20697-fcd5fb9.ngbuilds.io/.

mary-poppins commented:
You can preview 980cac4 at https://pr20697-980cac4.ngbuilds.io/.

mary-poppins commented:
You can preview c0bf4fb at https://pr20697-c0bf4fb.ngbuilds.io/.

mary-poppins commented:
You can preview beceb40 at https://pr20697-beceb40.ngbuilds.io/.

mary-poppins commented:
You can preview 0223d80 at https://pr20697-0223d80.ngbuilds.io/.

mary-poppins commented:
You can preview 170ea6a at https://pr20697-170ea6a.ngbuilds.io/.

mary-poppins commented:
You can preview 3249f09 at https://pr20697-3249f09.ngbuilds.io/.

mary-poppins commented:
You can preview a659d81 at https://pr20697-a659d81.ngbuilds.io/.

mary-poppins commented:
You can preview 23f7f3e at https://pr20697-23f7f3e.ngbuilds.io/.

wardbell commented:
Need someone to review and re-label

mary-poppins commented:
You can preview c586d20 at https://pr20697-c586d20.ngbuilds.io/.

shairez commented:
@wardbell Regarding [Testing without beforeEach()](https://pr20697-c586d20.ngbuilds.io/guide/testing#testing-without-beforeeach) -

I find that most of the time, in project I see, services often have more than one dependencies, and tests break whenever you change the order of the constructor dependencies.

That's basically the benefit of using the TestBed providers way, the injector makes the order of the params a non issue.

So I don't know if I would recommend not using beforeEach + TestBed as a first / best practice
That's my POV at least

wardbell commented:
@shairez I understand the "order of args" argument and I have broken myself more than once by shuffling the order (my OCD makes me want the in alpha order).

The counter argument is this.

1) Your tests almost always dead anyway if you add or subtract a param because ALL must be injected and your tests must provide them.
2) The other advantage of TestBed -- deep construction -- is rarely helpful here because you pretty much always write a test double for all of the service's test doubles.
3) The editor and compiler scream bloody murder if you screw up or change the param order. The tests immediately fail. So reordering params is not going to give you a sneaky bug.

That's why I think there is something compelling about this approach.

mary-poppins commented:
You can preview 81ab466 at https://pr20697-81ab466.ngbuilds.io/.

IgorMinar commented:
@StephenFluin or @vikerman would you have time to review this one from the technical correctness perspective.

@jenniferfell - just an fyi since this is a larger body of work that we'll need to review from the language/style perspective. We should discuss during the next docs meeting how we want to prioritize it.

shairez commented:
@wardbell 

Point 1 is valid, it won't solve it.

I use TestBed as an injector (now that the `ReflectiveInjector` is R.I.P)
Just to solve the re-ordering annoyance.

Use cases I see in large teams / code bases - 

You sometimes see components with 6 or more deps.
This begs for a refactor for a better design, but that's besides the point, let's say that you have repeating deps (Tracking, Logging, Translations, other than your services..)

Because you change the design frequently while you're building the feature, If you add a dep in the middle let's say, now you have to take a few seconds and see that you placed it in the right place on the test as well.

So the pain mostly comes for components with more than 1 or 2 deps, but more like 4 and above.

Plus, I think that in general, we should pick one way as a recommended way, say why it's recommended and place the rest of the options in the bottom of the doc as alternatives.

Less is more IMO, that's why I think it will help minimize the confusion for beginners who don't have the tools to judge which one is better, and also some of the veterans who comes from other frameworks / libraries and want to jump in and start testing. 

And in this particular case, I would go with a pattern that will hold up when deps grow and will reduce the unhappiness.

That's my opinion.

wardbell commented:
The merge conflicts on Jan 24 are due to the Stackblitz merge. I will resolve and move to stackblitz at same time.

wardbell commented:
Resolved merge conflicts introduced by StackBlitz PR
Both app and test stackblitzes are working
Docs are complete (no link problems) and passes lint checks.

Added `jasmine-marbles` to packages available in a stackblitz as this package appears in tests. Removed the package.json from `boilerplate/testing` because it was contributing nothing (I thought it would).

This PR is technically ready to merge. Awaiting reviewer approvals.

mary-poppins commented:
You can preview f01581a at https://pr20697-f01581a.ngbuilds.io/.

wardbell commented:
The CI failure (as of this comment) appears to be an unrelated saucelabs flak.

petebacondarwin commented:
Rebased and force pushed to see if I clear the Travis errors

mary-poppins commented:
You can preview 0c61a6c at https://pr20697-0c61a6c.ngbuilds.io/.

petebacondarwin commented on aio/tools/examples/shared/package.json:
```diff
@@ -36,7 +36,7 @@
     "core-js": "^2.4.1",
     "express": "^4.14.1",
-    "rxjs": "^5.5.0",
+    "rxjs": "^5.5.6",
```
I believe that you need to fix up the `yarn.lock` file associated with this `package.json`.

wardbell replied:
It would be nice too if I could remember why I bumped it. After reviewing RxJS changelog no reason leaps out at me. I'll roll back and confirm.

petebacondarwin commented on aio/src/app/embedded/code/code-example.component.ts:
```diff
@@ -48,7 +48,8 @@ export class CodeExampleComponent implements OnInit {
 
-    this.isAvoid = this.path.indexOf('.avoid.') !== -1;
+    const avoid = getBoolFromAttribute(element, 'avoid');
+    this.isAvoid = avoid || this.path.indexOf('.avoid.') !== -1;
```
No unit tests for this feature.

wardbell replied:
Fair enough. Will add.

wardbell replied:
Tests have been added.

mary-poppins commented:
You can preview 5f3dc46 at https://pr20697-5f3dc46.ngbuilds.io/.

petebacondarwin reviewed (changes requested):
This review is only from the point of view of AIO infrastructure, not the docs or examples themselves.

wardbell commented:
@petebacondarwin This works for master and should work for patch. Do you feel comfortable setting those labels? I'm not personally up-to-speed on that.

mary-poppins commented:
You can preview a324d1f at https://pr20697-a324d1f.ngbuilds.io/.

petebacondarwin commented:
I don't see any reason why this should not land on stable as well as master.
